pkg/builder/devcontainer.go

```diff
@@ -0,0 +1,384 @@
+	stdoutStr := string(result.StdOut)
+	stdoutStr = strings.TrimSuffix(stdoutStr, "\n")
+
+	// Find the index of the last newline character
+	lastNewline := strings.LastIndex(stdoutStr, "\n")
+
+	// If there is a newline character, slice the string from the index after the newline to the end
+	if lastNewline != -1 {
+		stdoutStr = stdoutStr[lastNewline+1:]
+	}
+
+	// Create a new Root object
+	root := &devcontainer.Root{}
+
+	// Unmarshal the JSON into the Root object
+	err = json.Unmarshal([]byte(stdoutStr), root)
+	if err != nil {
+		return err
+	}
+
+	b.postCreateCommands = append(b.postCreateCommands, root.MergedConfiguration.PostCreateCommands...)
+	b.postStartCommands = append(b.postStartCommands, root.MergedConfiguration.Entrypoints...)
+	b.postStartCommands = append(b.postStartCommands, root.MergedConfiguration.PostStartCommands...)
+
+	return nil
+}
+
+func (b *DevcontainerBuilder) startContainer() error {
+	ctx := context.Background()
+
+	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		return err
+	}
+
+	//	todo builder image from config
+	reader, err := cli.ImagePull(ctx, "daytonaio/workspace-project", image.PullOptions{
+		//	auth for builder image
+		//	RegistryAuth: ,
+	})
+	if err != nil {
+		return err
+	}
+	//	wait for pull to complete
+	_, err = io.Copy(io.Discard, reader)
+	if err != nil {
+		return err
+	}
+
+	dir := "/tmp/" + b.buildId + "/var/lib/docker"
+	err = os.MkdirAll(dir, 0755)
+	if err != nil {
+		return err
+	}
+
+	//	todo: mount folders
+	_, err = cli.ContainerCreate(ctx, &container.Config{
+		Image:      "daytonaio/workspace-project",
+		Entrypoint: []string{"sleep", "infinity"},
+	}, &container.HostConfig{
+		NetworkMode: "host",
+		Privileged:  true,
+		Binds: []string{
+			b.projectVolumePath + ":/project",
+			filepath.Dir(b.getLocalDockerSocket()) + ":/tmp/docker",
+			dir + ":/var/lib/docker",
+		},
+	}, nil, nil, b.buildId)
+	if err != nil {
+		return err
+	}
+
+	if err := cli.ContainerStart(ctx, b.buildId, container.StartOptions{}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (b *DevcontainerBuilder) startDocker() error {
+	ctx := context.Background()
+
+	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		return err
+	}
+
+	execConfig := types.ExecConfig{
+		AttachStdout: true,
+		AttachStderr: true,
+		Cmd:          []string{"dockerd", "-H", "unix:///tmp/docker/docker.sock", "-H", "unix:///var/run/docker.sock", "--insecure-registry", b.localContainerRegistryServer},
+	}
+
+	execResp, err := cli.ContainerExecCreate(ctx, b.buildId, execConfig)
+	if err != nil {
+		return err
+	}
+
+	execStartCheck := types.ExecStartCheck{
+		Detach: false,
+		Tty:    false,
+	}
+
+	attachResp, err := cli.ContainerExecAttach(ctx, execResp.ID, execStartCheck)
+	if err != nil {
+		return err
+	}
+	defer attachResp.Close()
+
+	go func() {
+		_, err = io.Copy(os.Stdout, attachResp.Reader)
```

Comment:
2 comments here:
1. We should use `dockerClient.ExecSync`
2. If we write to `os.Stdout`, the server logs might be bloated by this every time. I suggest you use `util.DebugLogWriter` to write this as debug logs.

Reply:
We can log all to the project logger for now, to let the user (client) see what's going on while booting up the builder.